### PR TITLE
[CLEANUP] Placed dependencies and plugins in parent pom

### PR DESCRIPTION
**Pull request by JoaoLMPereira**
dependencyManagement and pluginManagement. Placed and defined remaining
properties for dependencies, plugins and features.xml bundles versions
in parent pom.

### CODE REVIEW

**Comment by pamval** [comment]
@graimundo, please review

File: assemblies/karaf/features/src/main/resources/features.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
     <feature>cxf</feature>
     <feature>cxf-http-async</feature>
     <feature>war</feature>
-    <bundle>wrap:mvn:com.clearspring.analytics/stream/${clearspring.analytics-stream.version}</bundle>
+    <bundle>wrap:mvn:com.clearspring.analytics/stream/${clearspring.analytics.stream.version}</bundle>
```
Comment (graimundo) [inline assemblies/karaf/features/src/main/resources/features.xml]: Why the change from "-" to "." ?

File: assemblies/karaf/pom.xml [hunk 2]
```diff
@@ -63,7 +50,7 @@
   <build>
     <resources>
       <resource>
-        <directory>${project.basedir}/src/main/filtered-resources</directory>
```
Comment (graimundo) [inline assemblies/karaf/pom.xml]: Why the change in the name of the directory? Are the resources no longer filtered?

File: assemblies/karaf/pom.xml [hunk 3]
```diff
@@ -158,8 +144,7 @@
         </plugin>
       <plugin>
         <groupId>org.apache.karaf.tooling</groupId>
-        <artifactId>features-maven-plugin</artifactId>
-        <version>${karaf.version}</version>
+        <artifactId>karaf-maven-plugin</artifactId>
```
Comment (graimundo) [inline assemblies/karaf/pom.xml]: Did you change the plugin being used?

File: assemblies/karaf/features/src/main/resources/features.xml (hunk 1 above)
Comment (JoaoLMPereira) [inline assemblies/karaf/features/src/main/resources/features.xml]: I made that change to maintain consistency for properties names. The pattern I considered to give properties names is: {groupId or part of groupId}.{artifactId}.version

File: assemblies/karaf/src/main/descriptors/standalone-assembly.xml [hunk 4]
```diff
@@ -27,17 +27,25 @@
...
-      <directory>target/dependencies/apache-karaf-${karaf.version}</directory>
+      <directory>${project.build.directory}/dependencies/apache-karaf-${karaf.version}</directory>
       <outputDirectory>/apache-karaf-${karaf.version}</outputDirectory>
       <excludes>
         <exclude>etc/org.apache.karaf.features.cfg</exclude>
         <exclude>etc/startup.properties</exclude>
         <exclude>etc/custom.properties</exclude>
+        
+        <exclude>**/demos/**</exclude>
+        <exclude>LICENSE</exclude>
```
Comment (graimundo) [inline assemblies/karaf/src/main/descriptors/standalone-assembly.xml]: Why are you excluding license, readme, etc?

File: model/metricContributors/percentile/pom.xml [hunk 5]
```diff
@@ -6,7 +6,7 @@
     <groupId>pentaho</groupId>
     <artifactId>pentaho-data-profiling-model</artifactId>
     <version>6.0-SNAPSHOT</version>
-    <relativePath>../pom.xml</relativePath>
+    <relativePath>../../pom.xml</relativePath>
```
Comment (graimundo) [inline model/metricContributors/percentile/pom.xml]: The parent POM changed place?

File: assemblies/karaf/pom.xml (hunk 3 above)
Comment (JoaoLMPereira) [inline assemblies/karaf/pom.xml]: That is a modification that should have been sent with karaf 3.0.3 update commit, my bad

File: pom.xml [hunk 6]
```diff
@@ -42,30 +42,378 @@
...
 
     <!-- Other pentaho dependency versions -->
-    <kettle.version>6.0-SNAPSHOT</kettle.version>
-    <notification.version>6.0-SNAPSHOT</notification.version>
+    <kettle.version>${project.version}</kettle.version>
+    <notification.version>${project.version}</notification.version>
+    <pentaho-i18n-bundle.version>${project.version}</pentaho-i18n-bundle.version>
     <bean-matchers.version>0.9</bean-matchers.version>
+ 	
+    <source.jdk.version>1.6</source.jdk.version>
```
Comment (graimundo) [inline pom.xml]: Any idea why the source / target jdk is 1.6? Why is the project overriding the property from the parent POM?

File: assemblies/karaf/pom.xml (hunk 2 above)
Comment (JoaoLMPereira) [inline assemblies/karaf/pom.xml]: We do not use ${project.basedir}/src/main/filetered-resources folder, we use ${project.basedir}/src/main/resources and the files inside should be filtered

File: pom.xml [hunk 7]
```diff
@@ -42,30 +42,378 @@
...
   </properties>
 
-  <scm>
-    <connection>scm:git:git@github.com:${github.user}/${project.artifactId}.git</connection>
-    <url>scm:git:git@github.com:${github.user}/${project.artifactId}.git</url>
-    <developerConnection>scm:git:git@github.com:${github.user}/${project.artifactId}.git</developerConnection>
-    <tag>HEAD</tag>
-  </scm>
+  
+  <!-- copied from documentation pom.xml -->
```
Comment (graimundo) [inline pom.xml]: Why was this copied from the documentation POM? Do these profiles make sense on the other child POMs?

File: assemblies/karaf/src/main/descriptors/standalone-assembly.xml (hunk 4 above)
Comment (JoaoLMPereira) [inline assemblies/karaf/src/main/descriptors/standalone-assembly.xml]: We do not need them and the excluding is being done in pentaho-karaf-assembly project.

File: pom.xml [hunk 8]
```diff
@@ -42,30 +42,378 @@
...
+    <dependency>
+      <groupId>pentaho</groupId>
+      <artifactId>pentaho-data-profiling-api-core</artifactId>
+      <version>${project.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+      <dependency>
+      <groupId>pentaho</groupId>
+      <artifactId>pentaho-data-profiling-assemblies-karaf</artifactId>
```
Comment (graimundo) [inline pom.xml]: This dependency is not being used. Should be deleted.

File: model/metricContributors/percentile/pom.xml (hunk 5 above)
Comment (JoaoLMPereira) [inline model/metricContributors/percentile/pom.xml]: No, this modification was done because the relativePath was wrong. Before because we did not set the dependencyManagement, pluginManagement and properties in the project parent pom, it did not caused issues.

File: pom.xml [hunk 9]
```diff
@@ -42,30 +42,378 @@
...
+        </file>
+      </activation>
+      <properties>
+        <toolsjar>${java.home}/../Classes/classes.jar</toolsjar>
+      </properties>
+    </profile>
+  </profiles>
+  
+<dependencyManagement>
+  <dependencies>
```
Comment (graimundo) [inline pom.xml]: Validate that dependencies are actually used in child POMs. Remove them if they are not used.
If a dependency is only used by one of the child POMs then remove it from here and add it to that child POM.

File: pom.xml (hunk 7 above)
Comment (JoaoLMPereira) [inline pom.xml]: It sets the value for the toolsjar property. This property is being used in projects parent pom in the dependency com.sun.tools. For me it makes sense, because as far as I understood, the profiles purpose is dependent on the OS, so later other child POMs might use it.

File: api/core/pom.xml [hunk 10]
```diff
@@ -16,32 +16,24 @@
     <dependency>
       <groupId>com.google.code.bean-matchers</groupId>
       <artifactId>bean-matchers</artifactId>
-      <version>${bean-matchers.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline api/core/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: api/core/pom.xml [hunk 11]
```diff
@@ -16,32 +16,24 @@
...
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-api</artifactId>
-      <version>${slf4j.version}</version>
       <scope>provided</scope>
     </dependency>
     <dependency>
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-log4j12</artifactId>
-      <version>${slf4j.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline api/core/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: api/core/pom.xml [hunk 12]
```diff
@@ -16,32 +16,24 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline api/core/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: api/core/pom.xml [hunk 13]
```diff
@@ -50,14 +42,6 @@
...
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-jar-plugin</artifactId>
-        <version>2.5</version>
-        <executions>
-          <execution>
-            <goals>
-              <goal>test-jar</goal>
-            </goals>
-          </execution>
-        </executions>
```
Comment (brosander) [inline api/core/pom.xml]: @JoaoLMPereira we need to build a test jar for the other tests to resolve in

File: api/doc/rest/pom.xml [hunk 14]
```diff
@@ -16,44 +16,33 @@
...
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-api</artifactId>
-      <version>${slf4j.version}</version>
       <scope>provided</scope>
     </dependency>
     <dependency>
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-log4j12</artifactId>
-      <version>${slf4j.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline api/doc/rest/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: api/doc/rest/pom.xml [hunk 15]
```diff
@@ -16,44 +16,33 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline api/doc/rest/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: api/doc/rest/pom.xml [hunk 16]
```diff
@@ -16,44 +16,33 @@
...
     </dependency>
     <dependency>
       <groupId>pentaho</groupId>
       <artifactId>pentaho-data-profiling-api-core</artifactId>
-      <version>${project.version}</version>
       <classifier>tests</classifier>
-      <scope>test</scope>
```
Comment (brosander) [inline api/doc/rest/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: api/doc/rest/pom.xml [hunk 17]
```diff
@@ -16,44 +16,33 @@
...
     </dependency>
     <dependency>
       <groupId>com.google.code.bean-matchers</groupId>
       <artifactId>bean-matchers</artifactId>
-      <version>${bean-matchers.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline api/doc/rest/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: api/doc/rest/pom.xml [hunk 18]
```diff
@@ -62,14 +51,6 @@
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-jar-plugin</artifactId>
```
Comment (brosander) [inline api/doc/rest/pom.xml]: @JoaoLMPereira we need to build a test jar for the other tests to resolve in

File: api/json/pom.xml [hunk 19]
```diff
@@ -16,35 +16,28 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline api/json/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: api/osgi/pom.xml [hunk 20]
```diff
@@ -16,20 +16,15 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline api/osgi/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: assemblies/karaf/features/src/main/resources/features.xml [hunk 21]
```diff
@@ -20,11 +20,11 @@
     <bundle>mvn:pentaho/pentaho-data-profiling-model-core/${project.version}</bundle>
     <bundle>mvn:pentaho/pentaho-data-profiling-model-metrics/${project.version}</bundle>
     <bundle>mvn:pentaho/pentaho-data-profiling-model-metric-contributors-percentile/${project.version}</bundle>
-    <bundle>mvn:pentaho/pentaho-notification-webservice-bundle/${project.version}</bundle>
+    <bundle>mvn:pentaho/pentaho-notification-webservice-bundle/${notification.version}</bundle>
```
Comment (brosander) [inline assemblies/karaf/features/src/main/resources/features.xml]: @JoaoLMPereira currently the build team is working on replacing properties besides project.version but these changes will break the nightly until they get it working

File: api/json/pom.xml (hunk 19 above)
Comment (mattyb149) [inline api/json/pom.xml]: Why remove test scope for things like Mockito?

File: documentation/pom.xml [hunk 22]
```diff
@@ -12,67 +12,32 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline documentation/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: integration-tests/pom.xml [hunk 23]
```diff
@@ -19,99 +19,74 @@
...
       <groupId>pentaho</groupId>
       <artifactId>pentaho-data-profiling-api-core</artifactId>
-      <version>${project.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>pentaho</groupId>
       <artifactId>pentaho-data-profiling-model-core</artifactId>
-      <version>${project.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline integration-tests/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: kettle-integration/core/pom.xml [hunk 24]
```diff
@@ -22,61 +22,44 @@
...
     <dependency>
       <groupId>pentaho</groupId>
       <artifactId>pentaho-data-profiling-api-core</artifactId>
-      <version>${project.version}</version>
     </dependency>
     <dependency>
       <groupId>pentaho-kettle</groupId>
       <artifactId>kettle-ui-swt</artifactId>
-      <version>${kettle.version}</version>
-      <scope>provided</scope>
```
Comment (brosander) [inline kettle-integration/core/pom.xml]: @JoaoLMPereira this needs to be provided for the project to build in intellij (it has classes in the default package)

File: kettle-integration/core/pom.xml [hunk 25]
```diff
@@ -22,61 +22,44 @@
...
     </dependency>
     <dependency>
       <groupId>pentaho-kettle</groupId>
       <artifactId>kettle-engine</artifactId>
-      <version>${kettle.version}</version>
-      <scope>provided</scope>
```
Comment (brosander) [inline kettle-integration/core/pom.xml]: @JoaoLMPereira this needs to be provided for the project to build in intellij (it has classes in the default package)

File: kettle-integration/core/pom.xml [hunk 26]
```diff
@@ -22,61 +22,44 @@
...
     </dependency>
     <dependency>
       <groupId>pentaho-kettle</groupId>
       <artifactId>kettle-core</artifactId>
-      <version>${kettle.version}</version>
-      <scope>provided</scope>
```
Comment (brosander) [inline kettle-integration/core/pom.xml]: @JoaoLMPereira this needs to be provided for the project to build in intellij (it has classes in the default package)

File: kettle-integration/core/pom.xml [hunk 27]
```diff
@@ -22,61 +22,44 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.slf4j</groupId>
       <artifactId>slf4j-log4j12</artifactId>
-      <version>${slf4j.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline kettle-integration/core/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: kettle-integration/core/pom.xml [hunk 28]
```diff
@@ -22,61 +22,44 @@
...
     </dependency>
     <dependency>
       <groupId>com.google.code.bean-matchers</groupId>
       <artifactId>bean-matchers</artifactId>
-      <version>${bean-matchers.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline kettle-integration/core/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: kettle-integration/core/pom.xml [hunk 29]
```diff
@@ -22,61 +22,44 @@
...
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline kettle-integration/core/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: model/core/pom.xml [hunk 30]
```diff
@@ -16,44 +16,35 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline model/core/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: model/metricContributors/percentile/pom.xml [hunk 31]
```diff
@@ -17,48 +17,37 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline model/metricContributors/percentile/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: model/metricContributors/percentile/pom.xml [hunk 32]
```diff
@@ -17,48 +17,37 @@
...
     </dependency>
     <dependency>
       <groupId>pentaho</groupId>
       <artifactId>pentaho-data-profiling-api-core</artifactId>
       <classifier>tests</classifier>
-      <version>${project.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline model/metricContributors/percentile/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: model/metricContributors/percentile/pom.xml [hunk 33]
```diff
@@ -17,48 +17,37 @@
...
     </dependency>
     <dependency>
       <groupId>com.google.code.bean-matchers</groupId>
       <artifactId>bean-matchers</artifactId>
-      <version>${bean-matchers.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline model/metricContributors/percentile/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: model/metrics/pom.xml [hunk 34]
```diff
@@ -16,53 +16,41 @@
...
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-all</artifactId>
-      <version>${mockito.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline model/metrics/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: model/metrics/pom.xml [hunk 35]
```diff
@@ -16,53 +16,41 @@
...
     </dependency>
     <dependency>
       <groupId>pentaho</groupId>
       <artifactId>pentaho-data-profiling-api-core</artifactId>
       <classifier>tests</classifier>
-      <version>${project.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline model/metrics/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

File: model/metrics/pom.xml [hunk 36]
```diff
@@ -16,53 +16,41 @@
...
     </dependency>
     <dependency>
       <groupId>com.google.code.bean-matchers</groupId>
       <artifactId>bean-matchers</artifactId>
-      <version>${bean-matchers.version}</version>
-      <scope>test</scope>
```
Comment (brosander) [inline model/metrics/pom.xml]: @JoaoLMPereira this should probably stay a test dependency

**Comment by brosander** [comment]
@JoaoLMPereira added line comments

we use the test scope for dependencies (ex: mockito, bean matchers) we want to use while testing that are only needed by the test classes and should not be included at runtime

we need to build some test jars so that other modules can reuse test logic

provided scope is for dependencies that are to be looked for outside of karaf

${project.revision} is the only property that will correctly change from 6.0-SNAPSHOT to 6.0-NIGHTLY in feature files

there may be cases were we don't want to filter some resources, it's safer to have a separate folder that is obviously filtered so that people don't assume things with ${} remain as they are in the source file

**Comment by JoaoLMPereira** [comment]
Hello, thank you for the reviews so far. I am preparing a new commit, could you please review the new commit?
I pushed down the dependencies into dependencyManagement of project root pom. I did the same for the plugins, so the scope tests etc.. are described there. I have to change this to only push down shared dependencies and plugins. I am really sorry.

**Comment by brosander** [comment]
@JoaoLMPereira It's not a big deal. I just closed to prevent the merge button from being clicked accidentally and wanted to explain why :smile:.

**Comment by graimundo** [comment]
@brosander do you see any drawbacks on defining both the version and scope on the parent POM instead of having them scattered in the child POMs with the explicit ${x.version} variable? IMO having them defined in the parent POM and using Maven inheritance for the children makes the children POMs less cluttered.

Regarding the ${notification.version} and other properties, what @JoaoLMPereira did was to create intermediate variables that still point to ${project.version}. So if the build team changes that one only property it should still work. Nonetheless, I am not fully aware of how the build team overrides these values so please let me know if I am wrong.

@JoaoLMPereira I agree with @brosander on the filtered resource directory. Although the name is not standard maven directory layout, with it is easier to understand what resources are filtered and what are not. Besides we are using this naming on other pentaho projects.

**Comment by brosander** [comment]
@graimundo @JoaoLMPereira I should have looked more closely at the parent pom, I'd mistakenly assumed the dependencies were all being converted to compile, if all of the scope removals were just changes to use inheritance, that's fine with me, sorry about that.

I also didn't realize that notification.version had been changed to use project.version.

So besides the filtered-resources (which wasn't that high on my list of concerns) I'm good with the rest.  Sorry for not looking more closely first.

**Comment by graimundo** [comment]
No problem. Thank you for the feedback and looking into it :) @JoaoLMPereira will still need to check what should *really* go into the parent POM as opposed to have everything there ;)
